pkg/migrator/transformer.go

```diff
@@ -0,0 +1,341 @@
+
+	case inboxLogTableName:
+		data, ok := record.(*InboxLog)
+		if !ok {
+			return nil, fmt.Errorf("invalid record type: %T", record)
+		}
+
+		return t.TransformInboxLog(data)
+
+	case installationsTableName:
+		data, ok := record.(*Installation)
+		if !ok {
+			return nil, fmt.Errorf("invalid record type: %T", record)
+		}
+
+		return t.TransformInstallation(data)
+
+	case welcomeMessagesTableName:
+		data, ok := record.(*WelcomeMessage)
+		if !ok {
+			return nil, fmt.Errorf("invalid record type: %T", record)
+		}
+
+		return t.TransformWelcomeMessage(data)
+
+	default:
+		return nil, fmt.Errorf(
+			"Transform not implemented for table: %s",
+			record.TableName(),
+		)
+	}
+}
+
+// TransformGroupMessage converts GroupMessage to appropriate XMTPD envelope format.
+func (t *transformer) TransformGroupMessage(
+	groupMessage *GroupMessage,
+) (*envelopes.OriginatorEnvelope, error) {
+	if groupMessage == nil {
+		return nil, fmt.Errorf("groupMessage is nil")
+	}
+
+	if groupMessage.GroupID == nil {
+		return nil, fmt.Errorf("groupID is nil")
+	}
+
+	if len(groupMessage.Data) <= 0 {
+		return nil, fmt.Errorf("data is empty")
+	}
+
+	protoClientEnvelope := &proto.ClientEnvelope{
+		Payload: &proto.ClientEnvelope_GroupMessage{
+			GroupMessage: &mlsv1.GroupMessageInput{
+				Version: &mlsv1.GroupMessageInput_V1_{
+					V1: &mlsv1.GroupMessageInput_V1{
+						Data: groupMessage.Data,
+					},
+				},
+			},
+		},
+		Aad: &proto.AuthenticatedData{
+			TargetTopic: topic.NewTopic(topic.TOPIC_KIND_GROUP_MESSAGES_V1, groupMessage.GroupID[:]).
+				Bytes(),
+		},
+	}
+
+	return t.originatorEnvelope(
+		protoClientEnvelope,
+		groupMessageOriginatorID,
+		uint64(groupMessage.ID),
+	)
+}
+
+// TransformInboxLog converts InboxLog to appropriate XMTPD IdentityUpdate envelope format.
+func (t *transformer) TransformInboxLog(
+	inboxLog *InboxLog,
+) (*envelopes.OriginatorEnvelope, error) {
+	if inboxLog == nil {
+		return nil, fmt.Errorf("inboxLog is nil")
+	}
+
+	if inboxLog.InboxID == nil {
+		return nil, fmt.Errorf("inboxID is nil")
+	}
+
+	if len(inboxLog.IdentityUpdateProto) <= 0 {
+		return nil, fmt.Errorf("identityUpdateProto is empty")
+	}
+
+	var identityUpdateProto associations.IdentityUpdate
+
+	if err := protobuf.Unmarshal(inboxLog.IdentityUpdateProto, &identityUpdateProto); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal IdentityUpdateProto: %w", err)
+	}
+
+	protoClientEnvelope := &proto.ClientEnvelope{
+		Payload: &proto.ClientEnvelope_IdentityUpdate{
+			IdentityUpdate: &identityUpdateProto,
+		},
+		Aad: &proto.AuthenticatedData{
+			TargetTopic: topic.NewTopic(topic.TOPIC_KIND_IDENTITY_UPDATES_V1, inboxLog.InboxID[:]).
+				Bytes(),
+		},
+	}
+
+	return t.originatorEnvelope(
+		protoClientEnvelope,
+		inboxLogOriginatorID,
+		uint64(inboxLog.SequenceID),
+	)
+}
+
+// TransformInstallation converts Installation to appropriate XMTPD KeyPackage envelope format.
+func (t *transformer) TransformInstallation(
```

Comment:
In V3, when you upload a new key package we mutate the existing installation and change the key package field. That is going to be hard to make work in an ETL world. I would suggest we actually update V3 to have an append-only table of key packages joined to installations, which will be easier to migrate.

Reply:
Makes complete sense to me. New issue created to track the work. https://github.com/xmtp/xmtpd/issues/930